### PR TITLE
🚨🚨🚨 [`Refactor`] Move third-party related utility files into `integrations/` folder 🚨🚨🚨

### PR DESCRIPTION
**Pull request by younesbelkada**
# What does this PR do?

As per title and to address https://github.com/huggingface/transformers/pull/25077#discussion_r1283042392 
Let's move all third party libs (outside HF ecosystem) related utility files inside `lib_integrations/`, currently to the best of my knowledge bitsandbytes and deepspeed are the only 2 third party libs that we use as a plugin integration and is outside HF ecosystem

cc @sgugger and @stas00 as it touches DS related code

Let me see first if the CI passes

### CODE REVIEW

**Comment by stas00** [comment]
Hi Younes,

Unless I'm mistaken these would be BC breaking changes.

```
$ PYTHONPATH=src python -c "from transformers.deepspeed import HfDeepSpeedConfig"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ModuleNotFoundError: No module named 'transformers.deepspeed'
```


(In general for deepspeed integration issues please tag @pacman100 who is the current maintainer of the integration.)

**Comment by stas00** [comment]
unrelated - perhaps there could be a neater name than `lib_integrations` - this is awkward IMHO.

Perhaps just `integrations` or just `libs` or `external` - in other words one word would be smoother.

**Comment by sgugger** [comment]
Yes I had mentioned integrations as well as a name for the folder. Note that we usually do not guarantee backward compatibility with imports not at the init level (anything not imported in the main init or a subfolder init is considered private), but we can keep a deepspeed module that reimports `HfDeepSpeedConfig` if this line is in a lot of places.

**Comment by younesbelkada** [comment]
I did not used `integrations` as there is already a module named `integrations.py` in the same place. Maybe renaming `lib_integrations` to `external` would be better, WDYT?

**Comment by stas00** [comment]
Use a single `integration`? Since once you add the specifics it's a clear `transformers.integration.what`

as in ``transformers.integration.deepspeed`

**Comment by sgugger** [comment]
As long as the objects are re-imported, it's completely fine changing `integrations.py` to an `integrations` folder.

**Review by sgugger — COMMENTED** [review]
Not sur we need subfolders for DeepSpeed, bitandbytes and peft, they can jus be modules no?

**Comment by younesbelkada** [comment]
Ah yes makes sense, modified it accordingly

**Comment by younesbelkada** [comment]
Actually sorry, forgot to do it for peft, will do it now

**Comment by younesbelkada** [comment]
@sgugger the PR is now ready again ! 🙏 
The failing CI is related with the Hub rate limit issue we have been discussing internally (cc @ydshieh), would like to have a review if possible and I can put the PR on hold until the issue gets solved !

File: docs/source/en/main_classes/deepspeed.md
Comment (sgugger) [inline docs/source/en/main_classes/deepspeed.md:2068]:
```suggestion
The [`~integrations.HfDeepSpeedConfig`] is used to integrate Deepspeed into the 🤗 Transformers core
```
I think it's better to import directly from `integrations` since it's an important class. (Here and below)

File: src/transformers/modeling_utils.py
Comment (sgugger) [inline src/transformers/modeling_utils.py:664]: Same here and below.

**Review by sgugger — COMMENTED** [review]
Thanks for iterating! Let's import everything directly from `integrations`, but the rest LGTM!

**Review by pacman100 — COMMENTED** [review]
Thank you Younes for refactoring the integrations in a separate module 🤗.

As Stas mentioned, this is going to be a breaking change. For example, in `Accelerate` repo line https://github.com/huggingface/accelerate/blob/main/src/accelerate/utils/dataclasses.py#L728 needs to be updated. This would also mean updating the DeepSpeed documentation. Else the DeepSpeed integration of the trainer will break. 

Other than that, LGTM!

File: src/transformers/integrations/__init__.py
Comment (younesbelkada) [reply src/transformers/integrations/__init__.py:33]: Agreed!

**Comment by younesbelkada** [comment]
Thank you all! Feedback should be addressed, @pacman100 I am happy to propose a patch in accelerate after this gets merged, regarding the docs, please see: https://github.com/younesbelkada/transformers/blob/move-integrations/docs/source/en/main_classes/deepspeed.md which has been adapted with the refactor, let me know if there are some other place where the doc should be updated

**Comment by sgugger** [comment]
@younesbelkada as said before, we will need to keep the current DeepSpeed module to import back objects for backward compatibility (liek `file_utils.py` does), it's too breaking otherwise.

**Comment by younesbelkada** [comment]
Ah makes sense, reverted that and added a comment in the file

**Review by sgugger — COMMENTED** [review]
No the deepspeed module should just import objects and raise a deprecation warning (again, look at the `file_utils.py` module.

File: src/transformers/deepspeed.py
Comment (younesbelkada) [inline src/transformers/deepspeed.py]: Should this be a `FutureWarning` instead? I see most of deprecation warnings in transformers are `FutureWarning`, maybe because `DeprecationWarning` are not raised by default (from my tests locally)?

```suggestion
    FutureWarning,
```

**Review by younesbelkada — COMMENTED** [review]
Thanks and sorry for the multiple iterations, should be all good now, I left one question. However I couldn't find any deprecation warning in `file_utils.py`: https://github.com/huggingface/transformers/blob/main/src/transformers/file_utils.py neither on the main init, I also looked at the places where objects from that file being imported and could not see any depr warning as well

File: src/transformers/deepspeed.py
Comment (sgugger) [reply src/transformers/deepspeed.py]: Yes, it should be a FutureWarning.

File: src/transformers/__init__.py
Comment (sgugger) [inline src/transformers/__init__.py:118]: Missing an `integrations` key here.

**Review by sgugger — COMMENTED** [review]
Thanks! Do you know why the diff in the not_doctested file is so big?
As for the deprecation warning, I forgot we didn't put one in `file_utils` (probably cause it would be so breaking).

**Review by sgugger — APPROVED** [review]

File: src/transformers/deepspeed.py
Comment (younesbelkada) [reply src/transformers/deepspeed.py]: ok thanks, modified it in d50051a

File: src/transformers/__init__.py
Comment (younesbelkada) [reply src/transformers/__init__.py:118]: That key is already defined above: https://github.com/younesbelkada/transformers/blob/move-integrations/src/transformers/__init__.py#L108

**Review by pacman100 — APPROVED** [review]
Thank you @younesbelkada for iterating!

**Comment by younesbelkada** [comment]
Thanks everyone for your reviews, regarding the large diff on the doctest file my intuition is that it got large because I added a new line that changed the alphabetical order of the file but I am not sure. (maybe @ydshieh can confirm)
I also realised we might need to keep `utils/bitsandbytes.py` otherwise it would be too breaking, so I reverted it back and did a similar approach than `deepspeed.py`

**Comment by younesbelkada** [comment]
The large diff is fixed with https://github.com/huggingface/transformers/pull/25680

**Comment by younesbelkada** [comment]
@sgugger , ran the daily CI (with a reduced number of models tests) thanks to @ydshieh and you can see the results here: https://github.com/huggingface/transformers/actions/runs/5965296109 
compared the report with the report of the same day and I don't see any surprising difference / test failure caused by this PR. Therefore I think we can merge it ! Can you please have a final look ? 🙏 Thanks!

**Review by sgugger — APPROVED** [review]
Thanks for all the work on this!

**Pull request merged**
